peglib.h

```diff
@@ -971,66 +1080,112 @@ class LiteralString : public Ope
     void accept(Visitor& v) override;
 
     std::string lit_;
-	mutable bool init_is_word_;
-	mutable bool is_word_;
+    mutable bool init_is_word_;
+    mutable bool is_word_;
 };
 
 class CharacterClass : public Ope
     , public std::enable_shared_from_this<CharacterClass>
 {
 public:
-    CharacterClass(const std::string& chars) : chars_(chars) {}
+    CharacterClass(const std::string& chars) {
+
+        chars_ = chars;
+
+        auto i = 0u;
+        codepoint ch = 0;
+
+        while (i < chars.size()){
+            auto len = get_char(chars.c_str()+i, chars.size()-i, Encoding::UTF8, ch);
+            codepoints_.push_back(ch); // might push -1 but that's fine
+            i+=len;
+        }
+    }
 
     size_t parse(const char* s, size_t n, SemanticValues& sv, Context& c, any& dt) const override {
         c.trace("CharacterClass", s, n, sv, dt);
-        // TODO: UTF8 support
-        if (n < 1) {
-            c.set_error_pos(s);
-            return static_cast<size_t>(-1);
-        }
-        auto ch = s[0];
-        auto i = 0u;
-        while (i < chars_.size()) {
-            if (i + 2 < chars_.size() && chars_[i + 1] == '-') {
-                if (chars_[i] <= ch && ch <= chars_[i + 2]) {
-                    return 1;
+
+        if (c.encoding == Encoding::ASCII || c.encoding == Encoding::ANSI) {
+
+            if (n < 1) {
+                c.set_error_pos(s);
+                return static_cast<size_t>(-1);
+            }
+
+            auto ch = s[0];
+            auto i = 0u;
+            while (i < chars_.size()) {
+                if (i + 2 < chars_.size() && chars_[i + 1] == '-') {
+                    if (chars_[i] <= ch && ch <= chars_[i + 2]) {
+                        return 1;
+                    }
+                    i += 3;
+                } else {
+                    if (chars_[i] == ch) {
+                        return 1;
+                    }
+                    i += 1;
                 }
-                i += 3;
-            } else {
-                if (chars_[i] == ch) {
-                    return 1;
+            }
+        }
+        else if (c.encoding == Encoding::UTF8) {
+
+            codepoint ch = 0;
+            auto len = get_char(s, n, Encoding::UTF8, ch);
+
+            if (len < 1) {
+                c.set_error_pos(s);
+                return static_cast<size_t>(-1);
+            }
+
+            auto i = 0u;
+            while (i < codepoints_.size()) {
+                if (i + 2 < codepoints_.size() && codepoints_[i + 1] == '-') {
+                    if (codepoints_[i] != -1 && codepoints_[i] <= ch && ch <= codepoints_[i + 2]) {
+                        return len;
+                    }
+                    i += 3;
+                } else {
+                    if (codepoints_[i] == ch) {
+                        return len;
+                    }
+                    i += 1;
                 }
-                i += 1;
             }
         }
```

Comment:
This is the same code as above, except that the unicode array is used, not the char array. Also, the matched length might increase now.